conduit/pipeline.go

```diff
@@ -0,0 +1,286 @@
+	"github.com/algorand/indexer/importers"
+	"github.com/algorand/indexer/processors"
+)
+
+const autoLoadParameterConfigName = "conduit.yml"
+
+// PipelineConfig stores configuration specific to the conduit pipeline
+type PipelineConfig struct {
+	ConduitConfig *Config
+
+	PipelineLogLevel string `yaml:"LogLevel"`
+	// Store a local copy to access parent variables
+	Importer   map[string]interface{}   `yaml:"Importer"`
+	Processors []map[string]interface{} `yaml:"Processors"`
+	Exporter   map[string]interface{}   `yaml:"Exporter"`
+}
+
+// Valid validates pipeline config
+func (cfg *PipelineConfig) Valid() error {
+	if cfg.ConduitConfig == nil {
+		return fmt.Errorf("conduit configuration was nil")
+	}
+
+	if _, err := log.ParseLevel(cfg.PipelineLogLevel); err != nil {
+		return fmt.Errorf("pipeline log level (%s) was invalid: %w", cfg.PipelineLogLevel, err)
+	}
+
+	if len(cfg.Importer) == 0 {
+		return fmt.Errorf("importer configuration was empty")
+	}
+
+	if len(cfg.Exporter) == 0 {
+		return fmt.Errorf("exporter configuration was empty")
+	}
+
+	return nil
+}
+
+// MakePipelineConfig creates a pipeline configuration
+func MakePipelineConfig(logger *log.Logger, cfg *Config) (*PipelineConfig, error) {
+	if cfg == nil {
+		return nil, fmt.Errorf("empty conduit config")
+	}
+
+	// double check that it is valid
+	if err := cfg.Valid(); err != nil {
+		return nil, err
+	}
+
+	pCfg := PipelineConfig{PipelineLogLevel: logger.Level.String(), ConduitConfig: cfg}
+
+	// Search for pipeline configuration in data directory
+	autoloadParamConfigPath := filepath.Join(cfg.ConduitDataDir, autoLoadParameterConfigName)
+	paramConfigFound := util.FileExists(autoloadParamConfigPath)
+
+	if !paramConfigFound {
+		return nil, fmt.Errorf("could not find %s in data directory (%s)", autoLoadParameterConfigName, cfg.ConduitDataDir)
+	}
+
+	logger.Infof("Auto-loading Conduit Configuration: %s", autoloadParamConfigPath)
+
+	configStr, err := ioutil.ReadFile(autoloadParamConfigPath)
+	if err != nil {
+		return nil, err
+	}
+
+	err = yaml.Unmarshal(configStr, &pCfg)
+	if err != nil {
+		return nil, fmt.Errorf("config file (%s) was mal-formed yaml: %w", autoloadParamConfigPath, err)
+	}
+
+	if err := pCfg.Valid(); err != nil {
+		return nil, fmt.Errorf("config file (%s) had mal-formed schema: %w", autoloadParamConfigPath, err)
+	}
+
+	return &pCfg, nil
+
+}
+
+// Pipeline is a struct that orchestrates the entire
+// sequence of events, taking in importers, processors and
+// exporters and generating the result
+type Pipeline interface {
+	Start() error
+	Stop() error
+}
+
+type pipelineImpl struct {
+	ctx    context.Context
+	cfg    *PipelineConfig
+	logger *log.Logger
+
+	initProvider *data.InitProvider
+
+	importer   *importers.Importer
+	processors []*processors.Processor
+	exporter   *exporters.Exporter
+}
+
+func (p *pipelineImpl) Start() error {
+	p.logger.Infof("Starting Pipeline Initialization")
+
+	// TODO need to change config to map[string]interface{}
```

Comment:
done?

Reply:
Should read: "Need to change interfaces to accept config of map[string]interface{}